### ISSUE TITLE
Labels probability distribution using NeurASP and YOLO

### ISSUE THREAD

**Issue by Kmrs97** [azreasoners/NeurASP#3]
By using YOLOv5 with NeurASP, following your toy-car example, I noticed that, perhaps, due to the execution of the non_max_suppression() method, only one class/label has a positive probability (selects the maximum) and the rest are equal to 0. Τhis affects the stable models that are generated through NeurASP, and their ranking. Specifically, I would expect a behavior similar to the one shown in Section 3.2 of your IJCAI’20 paper, where, by applying state constraints, we get the most probable predictions that satisfy them, even though these predictions are not necessarily the ones with the maximum probability returned by the neural network. How can I get the stable models respecting the probability distribution returned by the Neural Network?

**Comment by zhunyoung** [comment]
Thank you for your interest in our work.

For each bounding box, the YOLO model (we used YOLOv3 but it should be similar to YOLOv5) doesn't output the probability distribution for all [80 labels](https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/master/data/coco.names) but only outputs a single probability along with a single label (and the coordinates of the bounding box). In this case, we only apply MAP (Maximum a posteriori) inference and only consider the labels of our interest `car`, `cat`, `person`, `truck`, and `other` as in section 3.1 and the following code.
https://github.com/azreasoners/NeurASP/blob/83df037b7229e1eccf06066bcc169a612a57067c/examples/toycar/infer.py#L15-L17
If YOLO predict, e.g., "this bounding box has 0.8 probability to be a dog", then we simply assign 0.8 to the label `other`. This is done by the last line below.
https://github.com/azreasoners/NeurASP/blob/83df037b7229e1eccf06066bcc169a612a57067c/examples/toycar/dataGen.py#L48-L51
If you want to prune out wrong predictions using constraints like what we did in section 3.2, you need to provide a neural network (or adjust YOLO if it's possible) so that it can produce the probabilities for all labels that are possibly correct. 

Please let me know if you have any questions and I'm happy to help.

**Comment by Kmrs97** [comment]
Thank you very much! Your response was very helpful!
